accounting/acct.proto

```diff
@@ -0,0 +1,147 @@
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// A protobuffer definition used for logging events reported to gNxI server.
+// This file assumes the following nomenclature:
+//
+//   Device >---connects---> gNxI accounting Server
+//
+// Device is a router, switch, PC, ... which connects to gNxI accounting server
+// to log accounting records.
+//
+// A gNxI accounting Server is a server which receives requests from devices and
+// logs them in the format specified using messages defined in this file.
+//
+// The current version covers event recording of shell/vendor-CLI and gRPC
+// based connections.
+
+syntax = "proto3";
+
+package gnsi.accounting;
+
+option go_package = "github.com/openconfig/gnsi/accounting";
+
+// Session identifier that identifies the context in which service
+// was executed.
+message SessionInfo {
+  // address of the device sending the event.
+  string device_address = 1;
+
+  // rest of the information for 5 tuple of the connection
+  uint32 layer4_proto = 2;
+  uint32 local_port = 3;
+  string remote_address = 4;
+  uint32 remote_port = 5;
+
+  // For multiplexing multiple channels over single 5 tuple
+  // - multiple CLI connections over single SSH
+  string channel_id = 6;
+}
+
+// Details of authentication - for gRPC or shell/vendor-CLI.
+message AuthenDetail {
+  // Identity string is used to identify the user that executed the
+  // command. For instance, it could be spiffe-id in the case of gRPC or
+  // unix-style user-name in the case of shell/vendor-CLI.
+  string identity = 1;
+
+  // Privilege level configured on the device.
+  uint32 privilege_level = 2;
+
+  // authentication status
+  enum AuthenStatus {
+    AUTHEN_STATUS_UNKNOWN = 0;
+    AUTHEN_STATUS_PERMIT = 1;
+    AUTHEN_STATUS_DENY = 2;
+  }
+  AuthenStatus status = 3;
+
+  // In case of STATUS_DENY, cause for the deny
+  string deny_cause = 4;
+}
+
+// Command details for shell/vendor-CLI
+message CommandService {
+  enum CmdServiceType {
+    CMD_SERVICE_UNKNOWN = 0;
+    CMD_SERVICE_SHELL = 1;
+    CMD_SERVICE_CLI = 2;
+  }
+  CmdServiceType service_type = 1;
+
+  // The command that was executed.
+  string cmd = 2;
+
+  // Arguments to a command above.
+  repeated string cmd_args = 3;
+}
+
+// Command details for openconfig gNxI commands.
+message GrpcService {
+  enum GrpcServiceType {
+    GRPC_SERVICE_UNKNWON = 0;
+    GRPC_SERVICE_GNMI = 1;
+    GRPC_SERVICE_GNOI = 2;
+    GRPC_SERVICE_GNSI = 3;
+    // GRIBI records could be very verbose.
+    // By default, the logging authorization
+    // should keep GRIBI logging off.
+    GRPC_SERVICE_GRIBI = 4;
+  }
+  GrpcServiceType service_type = 1;
+
```

Comment:
Do we want a `string` for service name here? Whilst gNMI, and gRIBI will likely have one -- I think gNSI and gNOI have separate services. If it's of interest that this was `gnoi.os` called to do an Install, then that would probably be something worth accounting for.

Equally, do we want to know what the RPC was -- since currently we'll just know that gNMI was called, not say, gNMI.Set and I'm not sure that it's possible to encode this in the `payload` below.

Reply:
Adding RPC seems good, yes. I was thinking that'd be in the payload, but perhaps that's optimistic :)

Adding a service name string also seems good... though if we have an enum of GrpcServiceType and then a rpc_name "Set" or "gnmi.Set" or "gnoi.os.Install" is that what you are meaning?

Reply:
Yes, I think your solution sounds good to me -- you'd have `service_type = GRPC_SERVICE_GNOI` and then you neatly managed the RPC by making it fully-qualified :-) i.e., `gnoi.os.Install()`.